### DISCUSSION TITLE
Watching changes on an array variable that is not registered to an input element only adds values, but does not remove them

### DISCUSSION THREAD

**Discussion by rvanpraet** [react-hook-form/react-hook-form#5681] [Q&A]
So I'm running into this strange behavior, I have some tables inside my form and you can move items from one table to another, updating the state of the tables. As I'm updating the state, I also want to set those state values in the form using:

setValue(fieldName, [ array values ])

I made sure to instantiate the form with default values for that specific fieldname with the values of said table, however I didn't register this fieldname to any actual input element because I want it's values to be managed by the tables. I do want to receive the correct values in my handleSubmit function.

I've reproduced the behavior in a small CSB: 

https://codesandbox.io/s/watched-array-values-issue-3qocn?file=/src/App.tsx

CSB shows the state array values and the watched values. As I add values to the array, both state and watched values update. However, when I make the array smaller (by pop, or filter in my application) only the state array changes, it seems setValue does not like it when I do this. Also, removing all items from the array and then adding them again makes the setValue add items from the beginning of the array.

Looking for some clarification :-)

**Comment by JamesWong1999** [comment]
can you share some steps on how to reproduce this issue?

**Reply by JamesWong1999** [reply]
i think i got the issue, `setValue` is only update inputs value, not reset the entire fields, perhaps you want to use `reset` isntead?

**Reply by JamesWong1999** [reply]
or if you want to delete item, maybe should use the `remove` action?

**Reply by rvanpraet** [reply]
Hi James, I figured it out. It's not about updating a useFieldArray, rather updating an array variable with setValue. I forgot to first register the 'fieldname' with register in a useEffect:

![afbeelding](https://user-images.githubusercontent.com/43333514/122777790-a9ce2780-d2ac-11eb-8040-f5278aa9bd57.png)

If anyone is interested in the original issue, steps to reproduce:

- Click 'Add Name' button -> A name is added to both the variable and the watched field
- Click 'Remove name' button -> You will see only the variable is updated correctly, which means setValue didn't correctly update the watched field

If you add the commented useEffect which first registers the field and then updates it, the steps above will produce the desired result.

**Reply by bluebill1049** [reply]
I will give a more detailed explanation today.

**Comment by bluebill1049** [comment]
Sorry for the late reply... it's been a very busy week with work and open source.

## What happens when input is not registered in `defaultValues`?

This is the explanation in the doc.
<img width="919" alt="Screen Shot 2021-06-22 at 9 41 55 am" src="https://user-images.githubusercontent.com/10513364/122840749-180efa80-d33e-11eb-98af-b158869372cf.png">

One of the behaviors for custom input is that the data can be `object`, `array`, or other data type. This is one of the nonstandard for a web form, which without proper `register` at the react hook form, we can't determine what type of inputs they are. 

```
defaultValues: {
  test: { firstName: 'bill', lastName: 'luo' },
  test1: ['test', 'test1']
},
```

without `register`

```
register('test.firstName')
register('test.lastName')
register('test1.0')
register('test1.1')
```

That probably will not be the case for some inputs that contain `array` or `object`.

to fix the incorrect registration:

```
register('test');
register('test1');
```

now react hook form know those are single/group input.

hope the above makes sense to everyone.
